tokio-timer/src/timer/registration.rs

```diff
@@ -0,0 +1,82 @@
+use Error;
+use timer::{Handle, Entry};
+
+use futures::Poll;
+
+use std::sync::Arc;
+use std::time::Instant;
+
+/// Registration with a timer.
+///
+/// The association between a `Sleep` instance and a timer is done lazily in
+/// `poll`
+#[derive(Debug)]
+pub(crate) struct Registration {
+    entry: Option<Arc<Entry>>,
+}
+
+impl Registration {
+    pub fn new(deadline: Instant) -> Registration {
+        match Handle::try_current() {
+            Ok(handle) => Registration::new_with_handle(deadline, handle),
```

Comment:
The handle is checked twice, once here in `new`, and then differently in `new_with_handle`. I don't know if the compiler sees them as the same...

Reply:
Sorry, I don't understand this comment.

Reply:
Basically, calling `Registration::new(deadline)` results in this:

```rust
let handle = match Handle::try_current() {
    Ok(handle) => (),
    Err(_) => return Registration::new_error(),
};
let inner = match handle.inner() {
    Some(inner) => inner,
    None => return Registration::new_error(),
};
```

Reply:
These do different things.

`Handle::try_current()` only checks the thread-local to pull the `Handle` for the current thread. `handle.inner()` tries to upgrade that `Weak<Inner>` -> `Arc<Inner>`.